src/prism.c

```diff
@@ -9770,6 +9770,17 @@ parser_lex(pm_parser_t *parser) {
             const uint8_t *ident_start = lex_mode->as.heredoc.ident_start;
             size_t ident_length = lex_mode->as.heredoc.ident_length;
 
+            // If we have a - or ~ heredoc, then we should adjust the start and the
+            // length of the ident so that we can match the lines.
+            if (
+                lex_mode->as.heredoc.indent == PM_HEREDOC_INDENT_TILDE ||
+                lex_mode->as.heredoc.indent == PM_HEREDOC_INDENT_DASH
+            ) {
+                size_t ident_leading_whitespace = pm_strspn_inline_whitespace(ident_start, (ptrdiff_t)ident_length);
+                ident_start += ident_leading_whitespace;
+                ident_length -= ident_leading_whitespace;
```

Comment:
I don't think this is going to work quite as we want. This is going to shrink the heredoc declaration such that only the main content is present, but that ignores the requirement that whitespace be present in the prefix.

So for example, with this PR, this is going to parse correctly:

```ruby
<<~'  FOO'
  hi
FOO
```

when it shouldn't.